backends/test/suite/__init__.py

```diff
@@ -0,0 +1,173 @@
+
+    COREML_TEST_FLOW = ("coreml", CoreMLTester)
+    ALL_TEST_FLOWS.append(COREML_TEST_FLOW)
+
+
+DTYPES = [
+    torch.int8,
+    torch.uint8,
+    torch.int16,
+    torch.uint16,
+    torch.int32,
+    torch.uint32,
+    torch.int64,
+    torch.uint64,
+    torch.float16,
+    torch.float32,
+    torch.float64,
+]
+
+FLOAT_DTYPES = [
+    torch.float16,
+    torch.float32,
+    torch.float64,
+]
+
+
+# The type of test function. This controls the test generation and expected signature.
+# Standard tests are run, as is. Dtype tests get a variant generated for each dtype and
+# take an additional dtype parameter.
+class TestType(Enum):
+    STANDARD = 1
+    DTYPE = 2
+
+
+# Function annotation for dtype tests. This instructs the test framework to run the test
+# for each supported dtype and to pass dtype as a test parameter.
+def dtype_test(func):
+    func.test_type = TestType.DTYPE
+    return func
+
+
+# Class annotation for operator tests. This triggers the test framework to register
+# the tests.
+def operator_test(cls):
+    _create_tests(cls)
+    return cls
+
+
+# Generate test cases for each backend flow.
+def _create_tests(cls):
+    for key in dir(cls):
+        if key.startswith("test_"):
+            _expand_test(cls, key)
+
+
+# Expand a test into variants for each registered flow.
+def _expand_test(cls, test_name: str):
+    test_func = getattr(cls, test_name)
+    for flow_name, tester_factory in ALL_TEST_FLOWS:
+        _create_test_for_backend(cls, test_func, flow_name, tester_factory)
+    delattr(cls, test_name)
+
+
+def _make_wrapped_test(test_func, *args, **kwargs):
+    def wrapped_test(self):
+        test_func(self, *args, **kwargs)
+
+    return wrapped_test
+
+
+def _make_wrapped_dtype_test(test_func, dtype, tester_factory):
+    def wrapped_test(self):
+        test_func(self, dtype, tester_factory)
+
+    return wrapped_test
+
+
+def _create_test_for_backend(
+    cls,
+    test_func: Callable,
+    flow_name: str,
+    tester_factory: Callable[[torch.nn.Module, Tuple[Any]], Tester],
+):
+    test_type = getattr(test_func, "test_type", TestType.STANDARD)
+
+    if test_type == TestType.STANDARD:
+        wrapped_test = _make_wrapped_test(test_func, tester_factory)
+        test_name = f"{test_func.__name__}_{flow_name}"
+        setattr(cls, test_name, wrapped_test)
+    elif test_type == TestType.DTYPE:
+        for dtype in DTYPES:
+            # wrapped_test = _make_wrapped_dtype_test(test_func, dtype, tester_factory)
+            wrapped_test = _make_wrapped_test(test_func, dtype, tester_factory)
+            dtype_name = str(dtype)[6:]  # strip "torch."
+            test_name = f"{test_func.__name__}_{dtype_name}_{flow_name}"
+            setattr(cls, test_name, wrapped_test)
+    else:
+        raise NotImplementedError(f"Unknown test type {test_type}.")
+
+
+class OperatorTest(unittest.TestCase):
+    def _test_op(self, model, inputs, tester_factory):
```

Comment:
static method?

Reply:
Sure. I'll update this up the stack, as there's a bit of refactoring on this with the reporting diffs.